### PR TITLE
Stateful model instance - infer synchronization issue fix

### PR DESCRIPTION
**Pull request by mzegla**

### CODE REVIEW

**Review by rasapala — APPROVED** [review]
LGTM

File: src/sequence.cpp [hunk 1]
```diff
@@ -25,6 +25,10 @@ void Sequence::updateLastActivityTime() {
     lastActivityTime = std::chrono::steady_clock::now();
 }
 
+const uint64_t& Sequence::getId() const {
```
Comment (dkalinowski) [inline src/sequence.cpp:28]: Uint64 is not a big overhead to copy. Maybe just return `uint64_t`?

Comment (dkalinowski) [reply src/sequence.cpp:28]: This also applies to all other functions that take and return const reference to uint64_t which I don't think is neccessary

File: src/sequence_manager.cpp [hunk 2]
```diff
@@ -79,67 +56,67 @@ Status SequenceManager::removeTimedOutSequences(std::chrono::steady_clock::time_
...
-    sequenceMutexPtr = sequences.at(sequenceId).getMutexPtr();
-    if (sequenceMutexPtr == nullptr)
-        return StatusCode::INTERNAL_ERROR;
-
     return StatusCode::OK;
 }
 
-Status SequenceManager::createSequence(const uint64_t& sequenceId, MutexPtr& sequenceMutexPtr) {
+Status SequenceManager::createSequence(const uint64_t& sequenceId) {
     /* TO DO: Generate unique ID if not provided by the client
```
Comment (dkalinowski) [inline src/sequence_manager.cpp:70]: Remove space from todo word so SDL checks can detect it before release

File: src/sequence_manager.cpp [hunk 3]
```diff
@@ -79,67 +56,67 @@ Status SequenceManager::removeTimedOutSequences(std::chrono::steady_clock::time_
...
-    auto status = addSequence(sequenceId);
-    if (!status.ok())
-        return status;
-
-    sequenceMutexPtr = sequences.at(sequenceId).getMutexPtr();
-    if (sequenceMutexPtr == nullptr)
-        return StatusCode::INTERNAL_ERROR;
-
+    if (sequenceExists(sequenceId)) {
+        spdlog::debug("Sequence with provided ID already exists");
```
Comment (dkalinowski) [inline src/sequence_manager.cpp:75]: Use SPDLOG_LOGGER_DEBUG

File: src/sequence_manager.cpp [hunk 4]
```diff
@@ -79,67 +56,67 @@ Status SequenceManager::removeTimedOutSequences(std::chrono::steady_clock::time_
...
+    } else {
+        spdlog::debug("Sequence with provided ID does not exists");
+        return StatusCode::SEQUENCE_MISSING;
+    }
     return StatusCode::OK;
 }
 
-Status SequenceManager::getSequenceMutexPtr(SequenceProcessingSpec& sequenceProcessingSpec, MutexPtr& sequenceMutexPtr) {
+Status SequenceManager::processRequestedSpec(SequenceProcessingSpec& sequenceProcessingSpec) {
     const uint32_t& sequenceControlInput = sequenceProcessingSpec.getSequenceControlInput();
```
Comment (dkalinowski) [inline src/sequence_manager.cpp:109]: I don't see a point why we use reference instead of copy here

File: src/statefulmodelinstance.cpp [hunk 5]
```diff
@@ -107,13 +107,16 @@ Status StatefulModelInstance::infer(const tensorflow::serving::PredictRequest* r
...
         return status;
 
-    MutexPtr sequenceMutexPtr = nullptr;
     std::unique_lock<std::mutex> sequenceManagerLock(sequenceManager->getMutex());
-    status = sequenceManager->getSequenceMutexPtr(sequenceProcessingSpec, sequenceMutexPtr);
+    status = sequenceManager->processRequestedSpec(sequenceProcessingSpec);
     if (!status.ok())
         return status;
+    const uint64_t& sequenceId = sequenceProcessingSpec.getSequenceId();
+    if (!sequenceManager->sequenceExists(sequenceId))
```
Comment (dkalinowski) [inline src/statefulmodelinstance.cpp:115]: We have 2 methods that do the same - `sequenceExists` and `hasSequence`. In HasSequence you return INTERNAL_ERROR if sequence is missing, why not to use it here? And get rid of one of the methods.

Comment (mzegla) [reply src/statefulmodelinstance.cpp:115]: sequenceExists is the low level method to just check if element exists in the sequences container (returning true or false).

hasSequence has more context and returns appropriate status depending on sequence presence in sequences container as well as its termination status (if the last request of the sequence is being processed, the sequence is not available for any other requests).

**Review by dkalinowski — APPROVED** [review]

**Pull request merged**